File: tests/lua/values_occ.lua
do
  local function Stream(a) return { __tag = "Stream", a } end
  local _print = print
  local to_string = tostring
  local function sum_squares(xs)
    local tmp = xs[1]
    local go = tmp._1
    return Stream({
      _1 = function(st)
        local tmp0 = go(st)
        local x = tmp0._1
        return { _2 = tmp0._2, _1 = x * x }
      end,
      _2 = tmp._2
    })
  end
  _print(to_string(sum_squares))
end

File: tests/lua/values_occ.ml
external val ( * ) : int -> int -> int = "function(x, y) return x * y end"

external val print : string -> unit = "print"
external val to_string : 'a -> string = "tostring"

type step 's 'a =
  | Yield of 'a * 's

type stream 'a =
  | Stream : forall 's. ('s -> step 's 'a) * 's -> stream 'a

let map f (Stream (go, start)) =
  let loop st =
    match go st with
    | Yield (x, s) -> Yield (f x, s)
  Stream (loop, start)

let sum_squares xs =
  map (fun x -> x * x) xs

let () =
  print (to_string sum_squares)
